### ISSUE TITLE
different behaviors in kl_div between torch.compile mode and eager mode

### ISSUE THREAD

**Issue by zoux1a** [pytorch/pytorch#117034]
### 🐛 Describe the bug

the output of `torch.nn.functional.kl_div` in torch.compile mode differs from eager mode. Above is a .pkl file which includes a specific input leading to this issue. I tried to generate `target` randomly and don't encountered the error.
[inputs.pkl.zip](https://github.com/pytorch/pytorch/files/13874497/inputs.pkl.zip)

to reproduce:
```python
import torch
import torch.nn as nn
import os, pickle


with open(os.path.join('/home/exp_data/torch-2','inputs.pkl'), 'rb') as f:
    inputs = pickle.load(f) # 'p1': ..., 'p2': ...


class Model(nn.Module):
    def __init__(self):
        super().__init__()
    def forward(self, input):  
        return torch.nn.functional.kl_div(input=input, reduction='none', target=inputs[0]['target'])        

x = torch.rand([1, 7, 3, 9, 9, 7], dtype=torch.float32)


model = Model().to(torch.device('cpu'))
eag = model(x)
opt = torch.compile(model.forward, mode='default')(x)

same_val = torch.allclose(eag.to('cpu'), 
                            opt.to('cpu'), 
                            rtol=1e-3, atol=1e-3, 
                            equal_nan=True)
if same_val == False : 
        raise ValueError('diff value')
```

### Versions

Collecting environment information...
PyTorch version: 2.3.0.dev20240106+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.9.17 (main, Jul 5 2023, 20:41:20) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-86-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 2070
GPU 1: NVIDIA GeForce RTX 2070
GPU 2: NVIDIA GeForce RTX 2070
GPU 3: NVIDIA GeForce RTX 2070

Nvidia driver version: 535.104.12
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 46 bits physical, 48 bits virtual
Byte Order: Little Endian
CPU(s): 32
On-line CPU(s) list: 0-31
Vendor ID: GenuineIntel
Model name: Intel(R) Xeon(R) CPU E5-2630 v3 @ 2.40GHz
CPU family: 6
Model: 63
Thread(s) per core: 2
Core(s) per socket: 8
Socket(s): 2
Stepping: 2
CPU max MHz: 3200.0000
CPU min MHz: 1200.0000
BogoMIPS: 4794.64
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm xsaveopt cqm_llc cqm_occup_llc dtherm ida arat pln pts md_clear flush_l1d
Virtualization: VT-x
L1d cache: 512 KiB (16 instances)
L1i cache: 512 KiB (16 instances)
L2 cache: 4 MiB (16 instances)
L3 cache: 40 MiB (2 instances)
NUMA node(s): 2
NUMA node0 CPU(s): 0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30
NUMA node1 CPU(s): 1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit: KVM: Mitigation: VMX disabled
Vulnerability L1tf: Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds: Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown: Mitigation; PTI
Vulnerability Mmio stale data: Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] pytorch-triton==2.2.0+e28a256d71
[pip3] torch==2.3.0.dev20240106+cu118
[pip3] torchaudio==2.2.0.dev20240106+cu118
[pip3] torchvision==0.18.0.dev20240106+cu118
[conda] cudatoolkit 11.8.0 h6a678d5_0 defaults
[conda] numpy 1.24.3 pypi_0 pypi
[conda] pytorch-triton 2.2.0+e28a256d71 pypi_0 pypi
[conda] torch 2.3.0.dev20240106+cu118 pypi_0 pypi
[conda] torchaudio 2.2.0.dev20240106+cu118 pypi_0 pypi
[conda] torchvision 0.18.0.dev20240106+cu118 pypi_0 pypi

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler

**Comment by alanhe151220037** [comment]
This difference is due to that the `target` is fp16 but `input` is fp32.

pytorch/aten/src/ATen/native/Loss.cpp:
```
Tensor kl_div(const Tensor& input, const Tensor& target, int64_t reduction, bool log_target) {
    ...
    output = at::xlogy(target, target) - target * input;
    ...
}
```
graph cache:
```
cpp_fused_mul_sub_xlogy_0 = async_compile.cpp('''
        ...
            auto tmp1 = c10::convert<float>(tmp0);
            auto tmp2 = std::isnan(tmp1);
            auto tmp3 = static_cast<float>(0.0);
            auto tmp4 = tmp1 == tmp3;
            auto tmp5 = std::log(tmp1);
            auto tmp6 = decltype(tmp1)(tmp1 * tmp5);
        ...
''')
...
def call(args):
    ...
    cpp_fused_mul_sub_xlogy_0(c_void_p(arg0_1.data_ptr()), c_void_p(arg1_1.data_ptr()), c_void_p(buf0.data_ptr()))
    ...
```
In eager mode, the `xlogy` here is called by `target` with dtype fp16 separately. But in compile mode, considering the dtype of `input` is fp32, the graph cache first cast `target` to fp32 (see `tmp1`) and then do the compute.

**Comment by bdhirsh** [comment]
Hey @zoux1a - I can run your repro, but I noticed that even in eager mode, the output of your model contains `nan`. Comparing nan with nan always returns false, which is why I see `torch.allclose()` returning false:


Adding this to your repro:
```
print(torch.any(torch.isnan(x)))
print(torch.any(torch.isnan(eag)))
print(torch.any(torch.isnan(opt)))
```

gives:
```
tensor(False)
tensor(True)
tensor(True)
```

**Comment by bdhirsh** [comment]
Sorry about that, didn't notice the `equal_nan=True` call in your repro.

This does not repro for me with `backend="aot_eager"` or `backend="aot_eager_decomp_partition", but it does repro with `backend="inductor"`. I also confirmed that it repros with both cpu and cuda inputs. Tentatively marking hi-pri

**Comment by peterbell10** [comment]
I've confirmed that masking out the infinities in eager, or promoting the eager `target` to float32 returns the same result as inductor. This suggests there is no inductor bug here, simply the intentional behavior that fused kernels don't truncate intermediates to half precision.

**Comment by masnesral** [comment]
Looks like we missed marking this issue as triaged, so doing that now.

**Comment by masnesral** [comment]
> This suggests there is no inductor bug here, simply the intentional behavior that fused kernels don't truncate intermediates to half precision.

Scrubbing old issues. Per this comment, there's nothing to do here. Please reopen of I've misunderstood the takeaway.
